src/Analyzer/Passes/FunctionToSubcolumnsPass.cpp

```diff
@@ -16,218 +17,419 @@
+        },
+    },
+    {
+        {TypeIndex::Nullable, "count"},
+        [](QueryTreeNodePtr &, FunctionNode & function_node, ColumnContext & ctx)
+        {
+            /// Replace `count(nullable_argument)` with `sum(not(nullable_argument.null))`
+            NameAndTypePair column{ctx.column.name + ".null", std::make_shared<DataTypeUInt8>()};
+            auto & function_arguments_nodes = function_node.getArguments().getNodes();
+
+            auto new_column_node = std::make_shared<ColumnNode>(column, ctx.column_source);
+            auto function_node_not = std::make_shared<FunctionNode>("not");
+
+            function_node_not->getArguments().getNodes().push_back(std::move(new_column_node));
+            resolveOrdinaryFunctionNodeByName(*function_node_not, "not", ctx.context);
+
+            function_arguments_nodes = {std::move(function_node_not)};
+            resolveAggregateFunctionNodeByName(function_node, "sum");
+        },
+    },
+    {
+        {TypeIndex::Nullable, "isNull"},
+        [](QueryTreeNodePtr & node, FunctionNode &, ColumnContext & ctx)
+        {
+            /// Replace `isNull(nullable_argument)` with `nullable_argument.null`
+            NameAndTypePair column{ctx.column.name + ".null", std::make_shared<DataTypeUInt8>()};
+            node = std::make_shared<ColumnNode>(column, ctx.column_source);
+        },
+    },
+    {
+        {TypeIndex::Nullable, "isNotNull"},
+        [](QueryTreeNodePtr &, FunctionNode & function_node, ColumnContext & ctx)
+        {
+            /// Replace `isNotNull(nullable_argument)` with `not(nullable_argument.null)`
+            NameAndTypePair column{ctx.column.name + ".null", std::make_shared<DataTypeUInt8>()};
+            auto & function_arguments_nodes = function_node.getArguments().getNodes();
+
+            function_arguments_nodes = {std::make_shared<ColumnNode>(column, ctx.column_source)};
+            resolveOrdinaryFunctionNodeByName(function_node, "not", ctx.context);
+        },
+    },
+    {
+        {TypeIndex::Tuple, "tupleElement"}, optimizeTupleOrVariantElement<DataTypeTuple>,
+    },
+    {
+        {TypeIndex::Variant, "variantElement"}, optimizeTupleOrVariantElement<DataTypeVariant>,
+    },
+};
+
+std::tuple<FunctionNode *, ColumnNode *, TableNode *> getTypedNodesForOptimization(const QueryTreeNodePtr & node)
+{
+    auto * function_node = node->as<FunctionNode>();
+    if (!function_node)
+        return {};
+
+    auto & function_arguments_nodes = function_node->getArguments().getNodes();
+    if (function_arguments_nodes.empty() || function_arguments_nodes.size() > 2)
+        return {};
+
+    auto * first_argument_column_node = function_arguments_nodes.front()->as<ColumnNode>();
+    if (!first_argument_column_node || first_argument_column_node->getColumnName() == "__grouping_set")
+        return {};
+
+    auto column_source = first_argument_column_node->getColumnSource();
+    auto * table_node = column_source->as<TableNode>();
```

Comment:
What if the source is a table function?

Reply:
Currently we don't optimize it and in most cases it doesn't make sense.